### DISCUSSION TITLE
Dynamic created blocks & widgets

### DISCUSSION THREAD

**Discussion by mvenghaus** [magewirephp/magewire#80] [Ideas]
Currently it is only possible to use magewire in layout xml files. I needed this for dynamic created blocks in templates and widgets. Here is my solution:

https://github.com/magewirephp/magewire/pull/79

**Comment by wpoortman** [comment]
OK, based on @mvenghaus MR #79, I've created a new branch https://github.com/magewirephp/magewire/tree/feature/dynamic-layout-v2. Within this branch, I've did a ton of experimenting on how the architecture could look like. I came up with a resolver structure where blocks can comply on one or multiple (lightweight) conditions. When true is returned, that resolver is used to construct and reconstruct the component.

- Construct stands for building the component based on the block during a preceding request (a.k.a page load).
- Reconstruct stands for rebuilding the component based on the request post-data during a subsequent request (a.k.a update)

Reconstruction will. be handed thanks to a new trait I've added to the base component. This trait can hold the used ResolverInterface. Thanks to this, we can glue a so called "public name" onto the request fingerprint. During a subsequent request, this public name will be used to search the related ResolverInterface and reconstruct's the component based on all the data coming along with the subsequent request post data.

**Files:**
- Magewirephp\Magewire\Model\Component\ResolverInterface
- Magewirephp\Magewire\Model\ComponentResolver

Components are being (re-) constructed at three particular points.
- During `to_html_before` (construct)
- During `to_html_after` (construct)
- During subsequent request (reconstruct)

Since Magewire initially only worked with Layout XML files, the Layout Resolver is the default fallback which can be changed if that would be desirable in the future. To support other types of components, like e.g. Widgets, we need a second Widget Resolver which can recognize Widgets based on the required Widget `\Magento\Widget\Block\BlockInterface` implementation.

Then, to make Widgets be picked up by Magewire, the Block data needs a 'magewire' parameter because this is the only gateway into Magewire. Making this any heavier would not benefit all the other blocks that are being loaded during a preceding request. To accomplice this, I came up with an idea.

1. First, we need a kind of basket class which holds multiple types of sub-baskets. These baskets can be named e.g. "widget" and should be extended from some kind of abstract AbstractSubBasket (don't use this name) class which can hold an array argument where DI can inject items into. A item can exist of a name and a Magewire Component instance (xsi:object). For widgets, this name should be identical to the widget ID (can be found in widget.xml). Now, we've mapped the widget to a specific Magewire component.
2. Make a after plugin on `\Magento\Widget\Model\Config\Converter` where we check inside the widget subbasket if the ID of the component exists. If so, a additional hidden field should be added to the widget data.

Manual test / Example:
![image](https://user-images.githubusercontent.com/5383956/228157984-cef4e67c-7b16-42f5-872b-3305c68b8eb4.png)
![image](https://user-images.githubusercontent.com/5383956/228157876-a9922646-0ef7-41c0-b56a-d6e781b9107f.png)

Why this way? Well, the widget ID is not used anywhere else in the widget process. So I've searched for a spot where I do still have the widget ID. Because the type is `text` and `visible` is set to `0`, this will become a hidden field which still holds a certain value, which needs to be the widget ID. This will result in `$block['data' => ['magewire' => 'visual_item']]`, which will pass the Magewire requirements since the Block only needs a `magewire` data key to pass.

3. Now during construction, we can create the component for this widget which will then be glued onto the Block by Magewire itself.

This sounds a bit complex, but in my eyes it's the only solid way to support multiple ways of constructing blocks with a Magewire component. Because it requires quite a lot of code, I think the widget resolver including all other logic should go into a separate `magewirephp/magewire-widgets` repository which depends on the Magewire core extension. This way developer can choose to use it only when needed and would skip not-required overhead.

Now my question to those interested: would this be a workable way without to much complexity. When implemented, it would only require one thing.

1. Glue a component instance to a widget block via the Widget basket based on the widget ID.
